lib/session_encryptor.rb

```diff
@@ -1,18 +1,148 @@
+  ]
+
   def load(value)
-    decrypted = encryptor.decrypt(value)
+    return LegacySessionEncryptor.new.load(value) if should_use_legacy_encryptor_for_read?(value)
+
+    _v2, ciphertext = value.split(':')
+    decrypted = outer_encryptor.decrypt(ciphertext)
+
+    session = JSON.parse(decrypted, quirks_mode: true).with_indifferent_access
+    kms_decrypt_sensitive_paths!(session)
 
-    JSON.parse(decrypted, quirks_mode: true).with_indifferent_access
+    session
   end
 
   def dump(value)
+    return LegacySessionEncryptor.new.dump(value) if should_use_legacy_encryptor_for_write?
+    value.deep_stringify_keys!
+
+    kms_encrypt_pii!(value)
+    kms_encrypt_sensitive_paths!(value, SENSITIVE_PATHS)
+    alert_or_raise_if_contains_sensitive_keys!(value)
     plain = JSON.generate(value, quirks_mode: true)
-    encryptor.encrypt(plain)
+    NEW_CIPHERTEXT_HEADER + ':' + outer_encryptor.encrypt(plain)
+  end
+
+  def kms_encrypt(text)
+    Base64.encode64(Encryption::KmsClient.new.encrypt(text, 'context' => 'session-encryption'))
+  end
+
+  def kms_decrypt(text)
+    Encryption::KmsClient.new.decrypt(
+      Base64.decode64(text), 'context' => 'session-encryption'
+    )
+  end
+
+  def outer_encryptor
+    Encryption::Encryptors::AttributeEncryptor.new
   end
 
   private
 
-  def encryptor
-    Encryption::Encryptors::SessionEncryptor.new
+  # The PII bundle is stored in the user session in the 'decrypted_pii' key.
+  # The PII is decrypted with the user's password when they successfully submit it and then
+  # stored in the session.  Before saving the session, this method encrypts the PII with KMS and
+  # stores it in the 'encrypted_pii' key.
+  #
+  # The PII is not frequently needed in its KMS-decrypted state. To reduce the
+  # risks around holding plaintext PII in memory during requests, this PII is KMS-decrypted
+  # on-demand by the Pii::Cacher.
+  def kms_encrypt_pii!(session)
+    return unless session.dig('warden.user.user.session', 'decrypted_pii')
+    decrypted_pii = session['warden.user.user.session'].delete('decrypted_pii')
+    session['warden.user.user.session']['encrypted_pii'] =
+      kms_encrypt(decrypted_pii)
+    nil
+  end
+
+  # This method extracts all of the sensitive paths that exist into a
+  # separate hash.  This separate hash is then encrypted and placed in the session.
+  # We use #reduce to build the nested empty hash if needed. If Hash#bury
+  # (https://bugs.ruby-lang.org/issues/11747) existed, we could use that instead.
+  def kms_encrypt_sensitive_paths!(session, sensitive_paths)
+    sensitive_data = {
+    }
+
+    sensitive_paths.each do |path|
+      all_but_last_key = path[0..-2]
+      last_key = path.last
+      value = session.dig(*all_but_last_key)&.delete(last_key)
+      if value
+        all_but_last_key.reduce(sensitive_data) do |hash, key|
+          hash[key] ||= {}
+
+          hash[key]
+        end
+
+        if all_but_last_key.blank?
+          sensitive_data[last_key] = value
```

Comment:
coverage

Reply:
good catch, there is a bug when the path is only one key deep, so I've fixed it and added a spec in 897da443f